### PR TITLE
Perform checks on matrix insertion in `MatrixCSR`

### PR DESCRIPTION
**Pull request by chrisrichardson**
Currently, there is no runtime check that matrix entries are correctly added (exist in sparsity), except in Debug mode.
The overhead to doing so is negligible, so this PR adds that check.

### CODE REVIEW

**Comment by jhale** [comment]
Looks good. I'd like to see more cheap sanity checks added throughout DOLFINx.

File: python/test/unit/la/test_matrix_csr.py [hunk 1]
```diff
@@ -215,3 +215,22 @@ def test_set_diagonal_distributed(dtype):
...
+
+@pytest.mark.parametrize('dtype', [np.float32, np.float64, np.complex64, np.complex128])
+def test_bad_entry(dtype):
+    sp = create_test_sparsity(6, 1)
+    mat1 = matrix_csr(sp, dtype=dtype)
+
+    # Set block in bs=1 matrix
+    with pytest.raises(RuntimeError):
+        mat1.set([1.0, 2.0, 3.0, 4.0], [0], [0], 2)
+    # Normal
```
Comment (garth-wells) [inline python/test/unit/la/test_matrix_csr.py:228]: What does this comment mean?

Comment (chrisrichardson) [reply python/test/unit/la/test_matrix_csr.py:228]: Data has block size but matrix does not (bs=1)

Comment (garth-wells) [reply python/test/unit/la/test_matrix_csr.py:228]: My comment was in invitation to improve the code comment :). `# Normal` carries no meaning.

Comment (chrisrichardson) [reply python/test/unit/la/test_matrix_csr.py:228]: Sure will do.

**Pull request merged**